clients/js/genjs.go

```diff
@@ -0,0 +1,249 @@
+	ServiceName string /* Service's internal name (aka the github repo/discovery name) */
+	Version     string
+	Description string
+	Methods     []string
+}
+
+var indexJSTmplStr = `
+const discovery = require("@clever/discovery");
+const request = require("request");
+const url = require("url");
+const opentracing = require("opentracing");
+
+// go-swagger treats handles/expects arrays in the query string to be a string of comma joined values
+// so...do that thing. It's worth noting that this has lots of issues ("what if my values have commas in them?")
+// but that's an issue with go-swagger
+function serializeQueryString(data) {
+  if (Array.isArray(data)) {
+    return data.join(",");
+  }
+  return data;
+}
+
+module.exports = class {{.ClassName}} {
+
+  address;
+  timeout;
+
+  constructor(options) {
+    options = options || {};
+
+    if (options.discovery) {
+      this.address = discovery("{{.ServiceName}}", "default").url();
+    } else if (options.address) {
+      this.address = options.address;
+    } else {
+      throw new Error("Cannot initialize {{.ServiceName}} without discovery or address");
+    }
+    if (options.timeout) {
+      this.timeout = options.timeout
+    }
+  }
+{{range $methodCode := .Methods}}{{$methodCode}}{{end}}
+}
+`
+
+var packageJSONTmplStr = `
+{
+  "name": "{{.PackageName}}",
+  "version": "{{.Version}}",
+  "description": "{{.Description}}",
+  "main": "index.js",
+  "dependencies": {
+    "@clever/discovery": "0.0.8",
+    "opentracing": "^0.11.1",
+    "request": "^2.75.0"
+  }
+}
+`
+
+var methodTmplStr = `
+  {{.MethodName}} = ({{range $param := .Params}}{{$param.JSName}}, {{end}}options, cb) => {
+    if (!cb && typeof options === "function") {
+      cb = options;
+      options = undefined;
+    }
+
+    if (!options) {
+      options = {};
+    }
+
+    const timeout = options.timeout || this.timeout;
+    const span = options.span;
+
+    const headers = {
+{{range $param := .HeaderParams}}
+        "{{$param.WagName}}": {{$param.JSName}},
+{{end}}
+    }
+
+    if (span) {
+      opentracing.inject(span, opentracing.FORMAT_TEXT_MAP, headers);
+      span.logEvent("{{.Method}} {{.Path}}");
+    }
+
+    const requestOptions = {
+      method: "{{.Method}}",
+      uri: this.address + "{{.BasePath}}{{.PathCode}}",
+      json: true,
+      timeout,
+      headers,
+      body: {
+{{range $param := .BodyParams}}
+        "{{$param.WagName}}": {{$param.JSName}},
+{{end}}
+      },
+      qs: {
+{{range $param := .QueryParams}}
+        "{{$param.WagName}}": serializeQueryString({{$param.JSName}}),
+{{end}}
+      },
+    };
+    return new Promise((resolve, reject) => {
+      const rejecter = (err) => {
+        reject(err);
+        if (cb) {
+          cb(err);
+        }
+      }
+      const resolver = (data) => {
+        resolve(data);
+        if (cb) {
+          cb(null, data);
+        }
+      }
+
+      request(requestOptions, (err, response, body) => {
```

Comment:
Yea, but using native promises directly made it easier to support callbacks and do the unmarshalling.